### PR TITLE
Adding a package UI - Add package name text field

### PR DESCRIPTION
**Pull request by bozidarsevo**
Part of #13551

## Description

- Added package name text field in "Custom" package tab when toggle "save this as a new package template" is ON.
- Added basic field validation, fields need to be non empty for buttons to be enabled.

Features remaining for the "Custom" tab:
- Saving package template logic
- Add package - gather all info from fields and using it for the package
- Use measure type that we got from backend (centimeters, inches etc.)

Features remaining for the "Carrier" tab:
- all

Features remaining for the "Saved" tab:
- all

Rest of the functionalities will be done in **separate** stacked PRs (done on top of this branch or trunk if this gets merged before).

## Testing information

1. Install and set up the Woo Shipping extension on your store.
2. Build and run the app with the `revampedShippingLabelCreation` feature flag enabled.
3. Create an order with the processing status and at least one physical product.
4. In the order details, select "Create Shipping Label."
5. Tap "Select a Package" button
6. Tap on the package type button to change the type
7. Tap in length, width and height fields to input values and check that the "Add Package" button is enabled when all 3 values are entered
8. Toggle "save this as a new package template" ON and "Save package template" button will show with text field for name above it.
9. Type some text in the text field
10. "Save package template" should be enabled when all dimension fields are non empty and text field for package name is non empty.

## Screenshots

https://github.com/user-attachments/assets/c7fc71f5-4b84-42b3-a4e2-aa290e947f89

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift [hunk 1]
```diff
@@ -1,5 +1,22 @@
 import SwiftUI
 
+final class WooShippingAddPackageViewModel: ObservableObject {
```
Comment (rachelmcr) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift]: Could we add unit tests for this view model? Not sure if you have that planned as a separate task.

Also, a small thing, but we usually add the view model as a separate file. Not a huge deal at this stage but it will likely help as we expand it to include more logic e.g. network calls.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift [hunk 2]
```diff
@@ -14,21 +31,39 @@ struct WooShippingAddPackageView: View {
...
-                return true
-            }
-        }
-        return fieldValues.count != WooShippingAddPackageDimensionView.DimensionType.allCases.count
-    }
+    // Holds selected package type when custom package is selected, it can be `box` or `envelope`
+    @State var packageType: PackageType = .box
+    // Holds value for toggle that determines if we are showing button for saving the template
+    @State var showSaveTemplate: Bool = false
+    @State var packageTemplateName: String = ""
```
Comment (rachelmcr) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift]: What do you think about moving these to the view model, as well? It could be helpful for unit testing, and for handling the values e.g. for `packageType` and `packageTemplateName` in network calls.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift [hunk 3]
```diff
@@ -64,100 +92,199 @@ struct WooShippingAddPackageView: View {
...
+                                }
+                            }
+                            .renderedIf(focusedField != nil)
                         }
                     }
+                    Toggle(isOn: $showSaveTemplate) {
+                        Text(Localization.saveNewPackageTemplate)
+                            .font(.subheadline)
+                    }
+                    .tint(Color(.withColorStudio(.wooCommercePurple, shade: .shade60)))
```
Comment (rachelmcr) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift]: I'd suggest using the semantic color `Color(.primary)` or `Color(.accent)` here (and elsewhere in this view). That makes it easier to adjust later if we update the app colors, and it adjusts the color in dark mode to fit our usual dark mode theme.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift [hunk 4]
```diff
@@ -64,100 +92,199 @@ struct WooShippingAddPackageView: View {
...
+    }
+
+    private func addPackageButtonTapped() {
+        // TODO: implement adding a package
+        guard validateCustomPackageInputFields() else { return }
+    }
+
+    private func savePackageAsTemplateButtonTapped() {
+        // TODO: implement saving package as a template
+        guard validateCustomPackageInputFields() else { return }
```
Comment (rachelmcr) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift]: I'd suggest putting actions like these in the view model, so we can unit test them.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift [hunk 5]
```diff
@@ -64,100 +92,199 @@ struct WooShippingAddPackageView: View {
...
-                HStack {
-                    // text
-                    Text(packageType.name)
-                        .bodyStyle()
-                    // arrows
-                    Spacer()
-                    Image(systemName: "chevron.up.chevron.down")
+                .padding(.horizontal)
+                .onChange(of: showSaveTemplate) { newValue in
+                    packageTemplateNameFieldFocused = newValue
```
Comment (rachelmcr) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift]: Do you think we should also reset `packageTemplateName` to an empty string when this is toggled? It feels a little odd that any text entered there is retained if I toggle this off and on again.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift [hunk 6]
```diff
@@ -64,100 +92,199 @@ struct WooShippingAddPackageView: View {
...
-            AdaptiveStack(spacing: 8) {
-                ForEach(WooShippingAddPackageDimensionView.DimensionType.allCases, id: \.self) { dimensionType in
-                    WooShippingAddPackageDimensionView(dimensionType: dimensionType, fieldValue: Binding(get: {
-                        return self.fieldValues[dimensionType] ?? ""
-                    }, set: { value in
-                        self.fieldValues[dimensionType] = value
-                    }))
+                .onChange(of: packageTemplateNameFieldFocused) { focused in
+                    if focused {
+                        DispatchQueue.main.asyncAfter(deadline: .now() + Constants.scrollToDelay, execute: {
```
Comment (rachelmcr) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift:174]: What's the reasoning behind the delay for this scroll action? It feels fine in testing, I'm just curious if it's needed for some reason or if it's just for the feel.

**Review by rachelmcr — COMMENTED** [review]
Looks good and work well in testing! I just have some suggestions below, primarily around what's included in the view model + unit testing.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift (hunk 5 above)
Comment (bozidarsevo) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift]: I can reset it but I am thinking what would be the least bad case:
- user enters a name, accidentally toggles it off, toggles it back on and the text is cleared (user needs to enter text again)
- user enters a name, toggles it off on purpose, toggles it back on and sees the last entered text (user can save with that name or change it) - **my preference**

I would prefer that the string is kept until the template is **saved**, and then we reset it. But we can see when we test it with more people what feels better?

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift (hunk 6 above)
Comment (bozidarsevo) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift:174]: When text field for name gets focused (packageTemplateNameFieldFocused changed) the SwiftUI automatically scrolls so that text field is visible, but the save button is not, and if I do not do the scrolling to button with a small delay this is basically called in paralel as scrolling to text field so save button is overlayed a bit with keyboard.
This is how it goes without a delay (half of the button is visible):

https://github.com/user-attachments/assets/a459a87e-a60c-45e9-a3ff-da6a8e70e81e

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift (hunk 2 above)
Comment (bozidarsevo) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift]: I was planning to but left it here in case I change something, but will move it! 👍

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift (hunk 5 above)
Comment (rachelmcr) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift]: Yeah, good point, that's a fair concern.

> But we can see when we test it with more people what feels better?

Sounds good!

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift (hunk 6 above)
Comment (rachelmcr) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddPackageView.swift:174]: Ah yeah, that makes sense. Thanks for the explanation! Maybe it's worth a small note in the code about that so that later on we remember why it's needed?

File: WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/WooShipping Create Shipping Labels/WooShippingAddCustomPackageViewModelTests.swift [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        XCTAssertEqual(viewModel.packageTemplateName, "")
+        XCTAssertEqual(viewModel.areFieldValuesInvalid, true)
+    }
+
+    func test_clear_field_values() {
+        // Given
+        let viewModel = WooShippingAddCustomPackageViewModel()
+
+        // When
+        viewModel.clearFieldValues()
```
Comment (rachelmcr) [inline WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/WooShipping Create Shipping Labels/WooShippingAddCustomPackageViewModelTests.swift:25]: Given that it inits with empty field values, how about entering some values (e.g. the `viewModel.fillWithDummyFieldValues()` method you use below) before clearing them in this test?

File: WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/WooShipping Create Shipping Labels/WooShippingAddCustomPackageViewModelTests.swift [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        XCTAssertEqual(viewModel.showSaveTemplate, false)
+        XCTAssertEqual(viewModel.packageTemplateName, "")
+    }
+
+    func test_it_with_not_all_field_values_set() {
+        // Given
+        let viewModel = WooShippingAddCustomPackageViewModel()
+
+        // When
+        viewModel.clearFieldValues()
```
Comment (rachelmcr) [inline WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/WooShipping Create Shipping Labels/WooShippingAddCustomPackageViewModelTests.swift]: Do you need this step in the test? If it inits with empty fields I'd think we could skip this and just do:

```
// When
viewModel.fieldValues[.height] = "1"
```

File: WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/WooShipping Create Shipping Labels/WooShippingAddCustomPackageViewModelTests.swift [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        XCTAssertEqual(viewModel.fieldValues.isEmpty, false)
+        XCTAssertEqual(viewModel.areFieldValuesInvalid, true)
+    }
+
+    func test_it_with_all_field_values_set() {
+        // Given
+        let viewModel = WooShippingAddCustomPackageViewModel()
+
+        // When
+        viewModel.clearFieldValues()
```
Comment (rachelmcr) [inline WooCommerce/WooCommerceTests/ViewRelated/Shipping Label/WooShipping Create Shipping Labels/WooShippingAddCustomPackageViewModelTests.swift]: Same as above, do you need this step in these tests?

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddCustomPackageViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+    func clearFieldValues() {
+        fieldValues.removeAll()
+    }
+
+    func resetValues() {
+        clearFieldValues()
+        showSaveTemplate = false
+        packageTemplateName = ""
+    }
```
Comment (rachelmcr) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/WooShipping Create Shipping Labels/WooShipping Package and Rate Selection/WooShippingAddCustomPackageViewModel.swift]: Should this also reset the `packageType` to `.box`? I could see it going either way (remembering the previously used package type or resetting it), but just asking because I noticed that if I select the envelope type that isn't reset.

**Review by rachelmcr — APPROVED** [review]
Looks great and continues to test well. 👍  I have a few minor questions below, mainly about the unit tests, but nothing blocking.

Also this is something to deal with later, maybe after checking with design, but I noticed that the screen isn't really usable in landscape mode. Just sharing it now since I just saw this:

Simulator Screenshot - iPhone 15 Pro - 2024-10-18 at 16 37 01

Simulator Screenshot - iPhone 15 Pro - 2024-10-18 at 16 39 02

**Comment by bozidarsevo** [comment]
> but I noticed that the screen isn't really usable in landscape mode

Since that screen is "heavy" with keyboard usage, we could disable landscape there. Since I do not see how we can adapt design to work on landscape, at least not on iPhones

**Comment by toupper** [comment]
> Since that screen is "heavy" with keyboard usage, we could disable landscape there. Since I do not see how we can adapt design to work on landscape, at least not on iPhones

We support landscape on every screen, so you can adapt it using a Scroll View. This is something we've done in other cases in the app.

**Comment by bozidarsevo** [comment]
> so you can adapt it using a Scroll View

I do use Scroll View here.

Currently the "Add Package" is not in the Scroll View since my understanding from slack thread was that that button needs to be on top of the keyboard (always).

I can move that button also in the Scroll View content but then that button will also scroll with all of the content, is that ok for us?

With that change (moving the Add Package button to Scroll View) it would be like this:

https://github.com/user-attachments/assets/ab090bbf-9dbc-45a2-97c4-057acaca84b2

**Comment by bozidarsevo** [comment]
> > so you can adapt it using a Scroll View
> 
> I do use Scroll View here.
> 
> Currently the "Add Package" is not in the Scroll View since my understanding from slack thread was that that button needs to be on top of the keyboard (always).
> 
> I can move that button also in the Scroll View content but then that button will also scroll with all of the content, is that ok for us?
> 
> With that change (moving the Add Package button to Scroll View) it would be like this:
> 
>  Simulator.Screen.Recording.-.iPhone.16.Pro.-.2024-10-21.at.11.04.20.mp4

I will do updates in a separate PR. It will not be just moving the button to scroll view, it will need few more adjustments since Spacer() behaves differently in scroll views etc.
But it will be fixed!
